<a href="https://colab.research.google.com/github/chineidu/NLP-Tutorial/blob/main/notebook/Projects/Data-mining/02-sentiments-analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sentiment Analysis

```sh
!pip install transformers[torch]
!pip install datasets evaluate
!pip install sentence-transformers pinecone-client
```

## Install Dependencies

In [2]:
!pip install transformers[torch]
!pip install datasets evaluate
!pip install sentence-transformers pinecone-client

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 2.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.4/179.4 kB 11.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 41.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 300.4/300.4 kB 38.3 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=32afb9b4991e164eb5c3af8f3417dac57649e2c6b6434059c3035a848e341e45
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence-transformers


In [45]:
# Built-in library
import re
import json
from typing import Any, Optional, TypeAlias, Union
import logging
import warnings

# Standard imports
import numpy as np
from pprint import pprint
import pandas as pd
import polars as pl
from rich import print
import torch

# Visualization
import matplotlib.pyplot as plt


# Pandas settings
pd.options.display.max_rows = 1_000
pd.options.display.max_columns = 1_000
pd.options.display.max_colwidth = 600

warnings.filterwarnings("ignore")

# Black code formatter (Optional)
# %load_ext lab_black

# auto reload imports
# %load_ext autoreload
# %autoreload 2

## Load And Prepare Dataset

- I'll be using a dataset containing approx. 90k hotel reviews by customers.
- The dataset can be loaded using HuggingFace datasets.

In [20]:
from datasets import load_dataset, Dataset


PATH: str = "ashraq/hotel-reviews"
reviews_data: Dataset = load_dataset(PATH, split="train")
reviews_data

Dataset({
    features: ['review_date', 'hotel_name', 'review'],
    num_rows: 93757
})

In [7]:
RANDOM_STATE: int = 123

print(reviews_data[:3])

{
    'review_date': ['8/3/2017', '8/3/2017', '8/3/2017'],
    'hotel_name': [
        'Park Plaza County Hall London',
        'Park Plaza County Hall London',
        'Park Plaza County Hall London'
    ],
    'review': [
        ' Extra bed was the worst breakfast queue was really terrible It s easy to tell people to come at a 
specific time though you have to arrange it somehow Parking is far away and hard to come back because of road 
diversions ',
        ' Just the location and view',
        ' Around the corner from the London eye and used it for a Trafalgar trip as the pickup for our tour was 
around the corner Walked down to the Borough market and took a river cruise to Greenwich Plent of eating spots 
close by as well Over the bridge was Westminster Abbey Parliament House and Big Ben and walked to Trafalgar Square 
as well also a laundry available which is very helpful '
    ]
}

In [21]:
Example: TypeAlias = dict[str, Any]

def get_text_length(example: Example) -> Example:
  """This returns the length of the review. """
  key: str = "review"
  result: Example = {"review_length": [len(x) for x in example.get(key)]}
  return result

In [22]:
reviews_data: Dataset = reviews_data.map(get_text_length, batched=True)
reviews_data

Map:   0%|          | 0/93757 [00:00<?, ? examples/s]

Dataset({
    features: ['review_date', 'hotel_name', 'review', 'review_length'],
    num_rows: 93757
})

In [31]:
# Cheeck the length of the reviews
N: int = 5
result: Example = reviews_data.sort("review_length", reverse=True)[:N]
print(result.get("review_length"))

[1966, 1959, 1943, 1937, 1937]

In [32]:
# Convert to tabular data
df: pd.DataFrame = reviews_data.to_pandas()

df_pl: pl.DataFrame = pl.from_pandas(data=df)
df_pl

shape: (93_757, 4)
┌─────────────┬───────────────────────────────┬───────────────────────────────────┬───────────────┐
│ review_date ┆ hotel_name                    ┆ review                            ┆ review_length │
│ ---         ┆ ---                           ┆ ---                               ┆ ---           │
│ str         ┆ str                           ┆ str                               ┆ i64           │
╞═════════════╪═══════════════════════════════╪═══════════════════════════════════╪═══════════════╡
│ 8/3/2017    ┆ Park Plaza County Hall London ┆  Extra bed was the worst breakfa… ┆ 220           │
│ 8/3/2017    ┆ Park Plaza County Hall London ┆  Just the location and view       ┆ 27            │
│ 8/3/2017    ┆ Park Plaza County Hall London ┆  Around the corner from the Lond… ┆ 384           │
│ 8/2/2017    ┆ Park Plaza County Hall London ┆  I wish you had wheat free snack… ┆ 33            │
│ …           ┆ …                             ┆ …                                 ┆ …             │
│ 8/4/2015    ┆ Holiday Inn London Kensington ┆  The room wasn t what was descri… ┆ 119           │
│ 8/4/2015    ┆ Holiday Inn London Kensington ┆  Location was good                ┆ 19            │
│ 8/4/2015    ┆ Holiday Inn London Kensington ┆  Bed room service was very late   ┆ 31            │
│ 8/4/2015    ┆ Holiday Inn London Kensington ┆  Great very easy to amend bookin… ┆ 102           │
└─────────────┴───────────────────────────────┴───────────────────────────────────┴───────────────┘

In [35]:
df_pl.describe(percentiles=[0.05, 0.10, 0.95])

shape: (10, 5)
┌────────────┬─────────────┬────────────┬────────┬───────────────┐
│ describe   ┆ review_date ┆ hotel_name ┆ review ┆ review_length │
│ ---        ┆ ---         ┆ ---        ┆ ---    ┆ ---           │
│ str        ┆ str         ┆ str        ┆ str    ┆ f64           │
╞════════════╪═════════════╪════════════╪════════╪═══════════════╡
│ count      ┆ 93757       ┆ 93757      ┆ 93757  ┆ 93757.0       │
│ null_count ┆ 0           ┆ 0          ┆ 0      ┆ 0.0           │
│ mean       ┆ null        ┆ null       ┆ null   ┆ 108.006496    │
│ std        ┆ null        ┆ null       ┆ null   ┆ 140.482189    │
│ …          ┆ …           ┆ …          ┆ …      ┆ …             │
│ median     ┆ null        ┆ null       ┆ null   ┆ 64.0          │
│ 5%         ┆ null        ┆ null       ┆ null   ┆ 9.0           │
│ 10%        ┆ null        ┆ null       ┆ null   ┆ 15.0          │
│ 95%        ┆ null        ┆ null       ┆ null   ┆ 342.0         │
└────────────┴─────────────┴────────────┴────────┴───────────────┘

In [41]:
# Convert back to HF dataset
reviews_data_1: Dataset = Dataset.from_pandas(df=df_pl.to_pandas())
reviews_data_1

Dataset({
    features: ['review_date', 'hotel_name', 'review', 'review_length'],
    num_rows: 88350
})

## Initialize Sentiment Analysis Model

- I'll be using a finetuned [RoBERTa](https://huggingface.co/cardiffnlp/twitter-roberta-base-sentiment-latest?text=This+hotel+is+not+so+great.+It+lacks+basic+facilities.) model.

In [55]:
# Set up the labels. The model label is obtained form the model card
label2id: dict[str, Any] = {'negative': 0, 'neutral': 1, 'positive': 2}
id2label: dict[str, Any] = {_id:_label for _label, _id in label2id.items()}

print(f"label2id: {label2id}")
print(f"id2label: {id2label}")

label2id: {'negative': 0, 'neutral': 1, 'positive': 2}

id2label: {0: 'negative', 1: 'neutral', 2: 'positive'}

In [56]:
from transformers import (AutoTokenizer,
                          AutoModelForSequenceClassification,
                          pipeline,)


MODEL_CHECKPOINT: str = "cardiffnlp/twitter-roberta-base-sentiment-latest"
TASK: str = "sentiment-analysis"

# Load the model from HuggingFace Hub
model = AutoModelForSequenceClassification.from_pretrained(MODEL_CHECKPOINT,
                                                           id2label=id2label,
                                                           label2id=label2id,)

# Load the tokenizer from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained(MODEL_CHECKPOINT)

Some weights of the model checkpoint at cardiffnlp/twitter-roberta-base-sentiment-latest were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [52]:
# Check if GPU is available
device: str = "cuda" if torch.cuda.is_available() else "cpu"

print(f"Device: {device}")

Device: cuda

Model labels: {'negative': 0, 'neutral': 1, 'positive': 2}

In [57]:
# Setup the classifier pipeline
clf = pipeline(task=TASK, model=model, device=device, tokenizer=tokenizer)

In [63]:
# Test the classifier pipeline
text: str = reviews_data[250].get("review")

print(text)

Everything was perfect especially the Italian restaurant where the food was outstanding Everyone were extremely 
nice and helpful and I mean everyone

In [64]:
clf(text)

[{'label': 'positive', 'score': 0.9884264469146729}]

<br><hr>

## Initialize Retriever

- **Retriever**:
  - In NLP, a `retriever` is a component that identifies and retrieves relevant documents or passages from a large corpus of text.

- Applications:
  - Question answering
  - Information retrieval
  - Text summarization

- `Sentence-transformer` model will be used as the `retriever`.

- I'll be using [all-MiniLM-L6-v2](sentence-transformers/all-MiniLM-L6-v2) retrieval because of its small size (***~90 mb***).
- For higher accuracy, you can use a more powerful transformer like [all-mpnet-base-v2](sentence-transformers/all-mpnet-base-v2).

In [67]:
from sentence_transformers import SentenceTransformer


RETRIEVER_CHECKPOINT: str = "sentence-transformers/all-MiniLM-L6-v2"
# Load the model from HuggingFace Hub
retriever = SentenceTransformer(RETRIEVER_CHECKPOINT,
                                device=device)

retriever

.gitattributes:   0%|          | 0.00/1.18k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

data_config.json:   0%|          | 0.00/39.3k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

train_script.py:   0%|          | 0.00/13.2k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Normalize()
)

In [75]:
retriever.get_max_seq_length(), retriever.get_sentence_embedding_dimension()

(256, 384)

### Initialize Pinecone Index

- I'll be using [Pinecone](https://www.pinecone.io/) as the vector DB.
- The dimension is `384` which was obtained by running:

```python
retriever.get_sentence_embedding_dimension()
# Returns: 384
```
- The metric used is `cosine` similarity.

In [78]:
import pinecone


YOUR_API_KEY: str = "YOUR_API_KEY"
YOUR_ENV: str = "us-west4-gcp-free"
DIMENSION: int = retriever.get_sentence_embedding_dimension()
METRIC : str = "cosine"
INDEX_NAME: str = "sentiment-analysis"

pinecone.init(api_key=YOUR_API_KEY, environment=YOUR_ENV)

In [80]:
# Check if index exists
if INDEX_NAME not in pinecone.list_indexes():
  pinecone.create_index(
      name=INDEX_NAME,
      dimension=DIMENSION,
      metric=METRIC
  )

# Connect to the created index
index = pinecone.Index(index_name=INDEX_NAME)

### Generate Embeddings

- Generate embeddings for all the reviews in the dataset.
- The embeddings, the sentiment label and score will be included in the Pinecone index as metadata.
- The metadata will be used to understand the customer opinions.